### ISSUE TITLE
systemd-journald is refusing valid coredumps with "Received file outside of allowed directories" when `Storage=journal` is used

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#27608]
### systemd version the issue has been seen with

latest main

### Used distribution

Arch Linux

### Linux kernel version used

`6.3.1-arch2-1`

### CPU architectures issue was seen on

None

### Component

systemd-coredump, systemd-journald

### Unexpected behaviour you saw

For the past week CentOS CI keeps trying to update the Arch Linux CI images but failing to do so, because of failing TEST-17-UDEV and TEST-74-AUX-UTILS. In both cases this is caused by systemd-journald refusing to accept a coredump from systemd-coredumpd. Also, it happens only sometimes:

TEST-74-AUX-UTILS:
```
[   25.264560] H systemd[1]: Started systemd-coredump@0-551-0.service.
...
[   25.332909] H (coredump)[552]: systemd-coredump@0-551-0.service: Executing: /usr/lib/systemd/systemd-coredump
[   25.356293] H testsuite-74.sh[548]: /tmp/make-dump: line 12:   549 Trace/breakpoint trap   (core dumped) "$bin" infinity
[   25.356500] H systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 67 (stored), closing.
[   25.491848] H systemd-journald[299]: Received file outside of allowed directories. Refusing.
[   25.357261] H testsuite-74.sh[539]: + /tmp/make-dump /tmp/test-dump SIGABRT
[   25.361145] H testsuite-74.sh[555]: + bin=/tmp/test-dump
[   25.361145] H testsuite-74.sh[555]: + sig=SIGABRT
[   25.361145] H testsuite-74.sh[555]: + ulimit -c unlimited
[   25.367608] H testsuite-74.sh[555]: + pid=556
[   25.367608] H testsuite-74.sh[555]: + sleep 1
[   25.368956] H systemd[1]: Received SIGCHLD from PID 552 (5).
[   25.369008] H testsuite-74.sh[556]: + /tmp/test-dump infinity
[   25.369628] H systemd[1]: Child 552 (5) died (code=exited, status=0/SUCCESS)
[   25.369671] H systemd[1]: systemd-coredump@0-551-0.service: Child 552 belongs to systemd-coredump@0-551-0.service.
[   25.369720] H systemd[1]: systemd-coredump@0-551-0.service: Main process exited, code=exited, status=0/SUCCESS (success)
[   25.369776] H systemd[1]: systemd-coredump@0-551-0.service: Deactivated successfully.
[   25.369823] H systemd[1]: systemd-coredump@0-551-0.service: Service will not restart (restart setting)
[   25.369875] H systemd[1]: systemd-coredump@0-551-0.service: Changed running -> dead
[   25.417042] H systemd[1]: systemd-coredump@0-551-0.service: bpf-lsm: Failed to delete cgroup entry from LSM BPF map: No such file or directory
[   25.417283] H systemd[1]: systemd-coredump@0-551-0.service: Consumed 7ms CPU time.
[   25.417330] H systemd[1]: Spawning thread to nuke /tmp/systemd-private-4fa4c27481c7418d94ff791cff33b21b-systemd-coredump@0-551-0.service-k9kwLm
[   25.417388] H systemd[1]: Spawning thread to nuke /var/tmp/systemd-private-4fa4c27481c7418d94ff791cff33b21b-systemd-coredump@0-551-0.service-3bKhIY
[   25.417429] H systemd[1]: Cannot stat /run/credentials/systemd-coredump@0-551-0.service: No such file or directory
[   25.417471] H systemd[1]: systemd-coredump@0-551-0.service: Collecting.
...
[   26.401220] H systemd[1]: Started systemd-coredump@1-561-0.service.
...
[   26.440860] H (coredump)[562]: systemd-coredump@1-561-0.service: Executing: /usr/lib/systemd/systemd-coredump
[   26.459774] H testsuite-74.sh[555]: /tmp/make-dump: line 12:   556 Aborted                 (core dumped) "$bin" infinity
[   26.595320] H systemd-journald[299]: Received file outside of allowed directories. Refusing.
[   26.460524] H testsuite-74.sh[539]: + mkdir -p /run/systemd/coredump.conf.d/
[   26.461436] H systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 67 (stored), closing.
[   26.466012] H testsuite-74.sh[539]: + printf '[Coredump]\nStorage=external'
[   26.466012] H testsuite-74.sh[539]: + /tmp/make-dump /tmp/test-dump SIGTRAP
[   26.469750] H testsuite-74.sh[566]: + bin=/tmp/test-dump
[   26.469750] H testsuite-74.sh[566]: + sig=SIGTRAP
[   26.469750] H testsuite-74.sh[566]: + ulimit -c unlimited
[   26.471853] H testsuite-74.sh[566]: + pid=567
[   26.471853] H testsuite-74.sh[566]: + sleep 1
[   26.473192] H testsuite-74.sh[567]: + /tmp/test-dump infinity
[   26.495439] H systemd[1]: Received SIGCHLD from PID 562 (5).
[   26.495526] H systemd[1]: Child 562 (5) died (code=exited, status=0/SUCCESS)
[   26.495570] H systemd[1]: systemd-coredump@1-561-0.service: Child 562 belongs to systemd-coredump@1-561-0.service.
[   26.495614] H systemd[1]: systemd-coredump@1-561-0.service: Main process exited, code=exited, status=0/SUCCESS (success)
[   26.495668] H systemd[1]: systemd-coredump@1-561-0.service: Deactivated successfully.
[   26.495740] H systemd[1]: systemd-coredump@1-561-0.service: Service will not restart (restart setting)
[   26.495784] H systemd[1]: systemd-coredump@1-561-0.service: Changed running -> dead
[   26.530464] H systemd[1]: systemd-coredump@1-561-0.service: bpf-lsm: Failed to delete cgroup entry from LSM BPF map: No such file or directory
[   26.530756] H systemd[1]: systemd-coredump@1-561-0.service: Consumed 83ms CPU time.
[   26.530818] H systemd[1]: Spawning thread to nuke /tmp/systemd-private-4fa4c27481c7418d94ff791cff33b21b-systemd-coredump@1-561-0.service-aPPhNu
[   26.530862] H systemd[1]: Spawning thread to nuke /var/tmp/systemd-private-4fa4c27481c7418d94ff791cff33b21b-systemd-coredump@1-561-0.service-NO2nOA
[   26.530958] H systemd[1]: Cannot stat /run/credentials/systemd-coredump@1-561-0.service: No such file or directory
[   26.531176] H systemd[1]: systemd-coredump@1-561-0.service: Collecting.
...
[   27.508072] H systemd[1]: Started systemd-coredump@2-573-0.service.
...
[   27.547603] H (coredump)[574]: systemd-coredump@2-573-0.service: Executing: /usr/lib/systemd/systemd-coredump
[   27.566526] H systemd-coredump[574]: [🡕] Process 567 (test-dump) of user 0 dumped core.
                                        
                                        Stack trace of thread 567:
                                        #0  0x00007f4cc9df9564 pause (libc.so.6 + 0xd4564)
                                        #1  0x0000562b9cacf356 n/a (/tmp/test-dump + 0x2356)
                                        #2  0x0000000000000000 n/a (n/a + 0x0)
                                        #3  0x4552555353455250 n/a (n/a + 0x0)
                                        ELF object binary architecture: AMD x86-64
[   27.567333] H testsuite-74.sh[566]: /tmp/make-dump: line 12:   567 Trace/breakpoint trap   (core dumped) "$bin" infinity
[   27.609417] H testsuite-74.sh[539]: + /tmp/make-dump /tmp/test-dump SIGABRT
[   27.609776] H systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 67 (stored), closing.
[   27.609951] H testsuite-74.sh[579]: + bin=/tmp/test-dump
[   27.609951] H testsuite-74.sh[579]: + sig=SIGABRT
[   27.609951] H testsuite-74.sh[579]: + ulimit -c unlimited
[   27.609951] H testsuite-74.sh[579]: + pid=580
[   27.609951] H testsuite-74.sh[579]: + sleep 1
[   27.613273] H systemd[1]: Received SIGCHLD from PID 574 (5).
[   27.613404] H testsuite-74.sh[580]: + /tmp/test-dump infinity
[   27.613978] H systemd[1]: Child 574 (5) died (code=exited, status=0/SUCCESS)
[   27.614034] H systemd[1]: systemd-coredump@2-573-0.service: Child 574 belongs to systemd-coredump@2-573-0.service.
[   27.614156] H systemd[1]: systemd-coredump@2-573-0.service: Main process exited, code=exited, status=0/SUCCESS (success)
[   27.614213] H systemd[1]: systemd-coredump@2-573-0.service: Deactivated successfully.
[   27.614263] H systemd[1]: systemd-coredump@2-573-0.service: Service will not restart (restart setting)
[   27.614311] H systemd[1]: systemd-coredump@2-573-0.service: Changed running -> dead
[   27.614359] H systemd[1]: systemd-coredump@2-573-0.service: bpf-lsm: Failed to delete cgroup entry from LSM BPF map: No such file or directory
[   27.614412] H systemd[1]: systemd-coredump@2-573-0.service: Consumed 56ms CPU time.
[   27.614464] H systemd[1]: Spawning thread to nuke /tmp/systemd-private-4fa4c27481c7418d94ff791cff33b21b-systemd-coredump@2-573-0.service-rYjZzc
[   27.614509] H systemd[1]: Spawning thread to nuke /var/tmp/systemd-private-4fa4c27481c7418d94ff791cff33b21b-systemd-coredump@2-573-0.service-VG4GeC
[   27.614555] H systemd[1]: Cannot stat /run/credentials/systemd-coredump@2-573-0.service: No such file or directory
[   27.614598] H systemd[1]: systemd-coredump@2-573-0.service: Collecting.
...
[   28.604332] H systemd[1]: Started systemd-coredump@3-585-0.service.
...
[   28.642045] H (coredump)[586]: systemd-coredump@3-585-0.service: Executing: /usr/lib/systemd/systemd-coredump
[   28.659725] H systemd-coredump[586]: [🡕] Process 580 (test-dump) of user 0 dumped core.
                                        
                                        Stack trace of thread 580:
                                        #0  0x00007fbf5f951564 pause (libc.so.6 + 0xd4564)
                                        #1  0x000055b5d616c356 n/a (/tmp/test-dump + 0x2356)
                                        #2  0x0000000000000000 n/a (n/a + 0x0)
                                        #3  0x4552555353455250 n/a (n/a + 0x0)
                                        ELF object binary architecture: AMD x86-64
[   28.660462] H testsuite-74.sh[579]: /tmp/make-dump: line 12:   580 Aborted                 (core dumped) "$bin" infinity
[   28.794178] H testsuite-74.sh[539]: + rm -fv /run/systemd/coredump.conf.d/99-external.conf
[   28.794560] H systemd[1]: Received SIGCHLD from PID 586 (5).
[   28.794788] H testsuite-74.sh[589]: removed '/run/systemd/coredump.conf.d/99-external.conf'
[   28.795460] H systemd[1]: Child 586 (5) died (code=exited, status=0/SUCCESS)
[   28.795514] H testsuite-74.sh[539]: + timeout 30 bash -c 'while [[ $(coredumpctl list -q --no-legend /tmp/test-dump | wc -l) -lt 4 ]]; do sleep 1; done'
[   28.796192] H systemd[1]: systemd-coredump@3-585-0.service: Child 586 belongs to systemd-coredump@3-585-0.service.
[   28.796248] H systemd[1]: systemd-coredump@3-585-0.service: Main process exited, code=exited, status=0/SUCCESS (success)
[   28.796300] H systemd[1]: systemd-coredump@3-585-0.service: Deactivated successfully.
[   28.796347] H systemd[1]: systemd-coredump@3-585-0.service: Service will not restart (restart setting)
[   28.796404] H systemd[1]: systemd-coredump@3-585-0.service: Changed running -> dead
[   28.796447] H systemd[1]: systemd-coredump@3-585-0.service: bpf-lsm: Failed to delete cgroup entry from LSM BPF map: No such file or directory
[   28.796499] H systemd[1]: systemd-coredump@3-585-0.service: Consumed 51ms CPU time.
[   28.796553] H systemd[1]: Spawning thread to nuke /tmp/systemd-private-4fa4c27481c7418d94ff791cff33b21b-systemd-coredump@3-585-0.service-oBiRFB
[   28.796599] H systemd[1]: Spawning thread to nuke /var/tmp/systemd-private-4fa4c27481c7418d94ff791cff33b21b-systemd-coredump@3-585-0.service-n8QxuM
[   28.796643] H systemd[1]: Cannot stat /run/credentials/systemd-coredump@3-585-0.service: No such file or directory
[   28.796820] H systemd[1]: systemd-coredump@3-585-0.service: Collecting.
```

Full journal: https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/vagrant-make-cache/23/artifact//systemd-centos-ci/artifacts_all/artifacts_qhzdm0gd/vagrant-logs.gsR/vagrant-arch-testsuite.fFH/TEST-74-AUX-UTILS/system.journal

TEST-17-UDEV:
```
[   20.246022] H systemd-udevd[595]: null: Worker [602] terminated by signal 6 (ABRT).
[   20.246484] H systemd[1]: Received SIGCHLD from PID 671 (5).
[   20.246564] H systemd[1]: Child 671 (5) died (code=exited, status=0/SUCCESS)
[   20.246615] H systemd[1]: systemd-coredump@0-670-0.service: Child 671 belongs to systemd-coredump@0-670-0.service.
[   20.380748] H systemd-journald[298]: Received file outside of allowed directories. Refusing.
[   20.246831] H systemd[1]: systemd-coredump@0-670-0.service: Main process exited, code=exited, status=0/SUCCESS (success)
[   20.246888] H systemd[1]: systemd-coredump@0-670-0.service: Deactivated successfully.
[   20.246932] H systemd[1]: systemd-coredump@0-670-0.service: Service will not restart (restart setting)
[   20.246974] H systemd[1]: systemd-coredump@0-670-0.service: Changed running -> dead
[   20.270426] H systemd[1]: systemd-coredump@0-670-0.service: bpf-lsm: Failed to delete cgroup entry from LSM BPF map: No such file or directory
[   20.270715] H systemd[1]: systemd-coredump@0-670-0.service: Consumed 115ms CPU time.
[   20.270777] H systemd[1]: Spawning thread to nuke /tmp/systemd-private-e2b120c74c654aaeae402f60bdb6fd39-systemd-coredump@0-670-0.service-S3Ft65
[   20.270849] H systemd[1]: Spawning thread to nuke /var/tmp/systemd-private-e2b120c74c654aaeae402f60bdb6fd39-systemd-coredump@0-670-0.service-zgc5Bv
[   20.270930] H systemd[1]: Cannot stat /run/credentials/systemd-coredump@0-670-0.service: No such file or directory
[   20.271132] H systemd[1]: systemd-coredump@0-670-0.service: Collecting.
```

Full journal: https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/vagrant-make-cache/23/artifact//systemd-centos-ci/artifacts_all/artifacts_qhzdm0gd/vagrant-logs.gsR/vagrant-arch-testsuite.fFH/TEST-17-UDEV/system.journal

I have no idea, so far, why it happens only in the updated image (probably something in kernel has changed?).

**Comment by mrc0mmand** [comment]
Interesting, after adding a dumb printf to see what's the "not allowed" directory I see:
```
[   53.189357] H (coredump)[447]: systemd-coredump@0-446-0.service: Executing: /usr/lib/systemd/systemd-coredump
[   53.355333] H systemd-journald[175]: server_process_native_file: k=/memfd:sd-systemd-coredum (deleted)
[   53.355760] H systemd-journald[175]: Received file outside of allowed directories. Refusing.
[   53.400381] H testsuite-74.sh[443]: /tmp/make-dump: line 12:   444 Trace/breakpoint trap   (core dumped) "$bin" infinity
[   53.408899] H testsuite-74.sh[434]: + /tmp/make-dump /tmp/test-dump SIGABRT
```

**Comment by mrc0mmand** [comment]
Ah, this is easily reproducible locally by telling coredumpd to store the coredumps in journal:

```
# mkdir /run/systemd/coredump.conf.d
# printf "[Coredump]\nStorage=journal\n" >/run/systemd/coredump.conf.d/99-journal.conf
# ulimit -c unlimited
# bash -c 'kill -SEGV $$'
Segmentation fault (core dumped)
# journalctl -o short-monotonic --no-pager -e -n 10
[ 4533.671550] arch.localdomain systemd[1]: Stopping systemd-hostnamed.service...
[ 4533.679273] arch.localdomain systemd[1]: systemd-hostnamed.service: Deactivated successfully.
[ 4533.709810] arch.localdomain systemd[1]: Stopped systemd-hostnamed.service.
[ 4533.733541] arch.localdomain systemd[1]: Starting systemd-hostnamed.service...
[ 4533.766400] arch.localdomain systemd[1]: Started systemd-hostnamed.service.
[ 4563.797908] arch.localdomain systemd[1]: systemd-hostnamed.service: Deactivated successfully.
[ 4601.111047] arch.localdomain systemd[1]: Created slice system-systemd\x2dcoredump.slice.
[ 4601.126877] arch.localdomain systemd[1]: Started systemd-coredump@0-507980-0.service.
[ 4601.199205] arch.localdomain systemd[1]: systemd-coredump@0-507980-0.service: Deactivated successfully.
[ 4601.310364] arch.localdomain systemd-journald[186]: Received file outside of allowed directories. Refusing.
```

**Comment by mrc0mmand** [comment]
> Which kernel version affects the issue?

6.3.1-arch2-1

**Comment by yuwata** [comment]
Yeah, I mean, how about older version?

**Comment by mrc0mmand** [comment]
Current CentOS CI has 6.2.13.arch1-1 which seems to work fine.

**Comment by mrc0mmand** [comment]
And the winner is c29715a8f77d96cd731b4a3083b3a852b3b61eb8 - after reverting it, the coredumps work once again.

/cc @t-8ch

**Comment by t-8ch** [comment]
I guess `memfd_set_sealed` and `memfd_get_sealed` need adaption to handle `F_SEAL_EXEC`.
This will also need code to handle kernels that don't know `F_SEAL_EXEC` and don't return it from `F_GET_SEALS` and don't accept it in `F_ADD_SEALS`.

**Comment by yuwata** [comment]
Fix is waiting in #27659.
